pineappl/src/boc.rs

```diff
@@ -209,6 +212,460 @@ impl Scales {
+            match range.end_bound().cloned() {
+                Bound::Excluded(end) => Bound::Included(end),
+                Bound::Included(end) => Bound::Included(end + 1),
+                Bound::Unbounded => Bound::Unbounded,
+            },
+        );
+
+        Self::new(
+            self.bins
+                .iter()
+                .enumerate()
+                .filter_map(|(index, bin)| {
+                    if range.contains(&index) {
+                        Some(bin.clone())
+                    } else {
+                        None
+                    }
+                })
+                .collect(),
+            self.fill_limits
+                .iter()
+                .enumerate()
+                .filter_map(|(index, &limit)| {
+                    if range_plus_one.contains(&index) {
+                        Some(limit)
+                    } else {
+                        None
+                    }
+                })
+                .collect(),
+        )
+    }
+
+    /// TODO
+    ///
+    /// # Errors
+    ///
+    /// TODO
+    // TODO: change range to `RangeBounds<usize>`
+    pub fn merge(&self, range: Range<usize>) -> Result<Self, ()> {
+        // TODO: allow more flexible merging
+        if !self
+            .slices()
+            .iter()
+            .any(|&Range { start, end }| (start <= range.start) && (range.end <= end))
+        {
+            // TODO: implement proper error handling
+            return Err(());
+        }
+
+        let mut limits: Vec<_> = self.bins.iter().map(|bin| bin.limits().to_vec()).collect();
+        let mut normalizations = self.normalizations();
+        let mut fill_limits = self.fill_limits.clone();
+        let dim = self.dimensions();
+
+        limits[range.start][dim - 1].1 = limits[range.end - 1][dim - 1].1;
+        normalizations[range.start] += normalizations[range.start + 1..range.end]
+            .iter()
+            .sum::<f64>();
+
+        limits.drain(range.start + 1..range.end);
+        normalizations.drain(range.start + 1..range.end);
+        fill_limits.drain(range.start + 1..range.end);
+
+        Self::new(
+            limits
+                .into_iter()
+                .zip(normalizations)
+                .map(|(limits, normalization)| Bin::new(limits, normalization))
+                .collect(),
+            fill_limits,
+        )
+    }
+
+    /// TODO
+    ///
+    /// # Panics
+    ///
+    /// TODO
+    pub fn remove(&mut self, index: usize) -> Bin {
+        assert!(self.len() > 1);
+
+        self.fill_limits.pop().unwrap();
+        self.bins.remove(index)
+    }
+
+    /// TODO
+    #[must_use]
+    pub fn bins_partial_eq_with_ulps(&self, other: &Self, ulps: i64) -> bool {
+        self.bins
+            .iter()
+            .zip(other.bins())
+            .all(|(lhs, rhs)| lhs.partial_eq_with_ulps(rhs, ulps))
+    }
+}
+
+impl FromStr for BinsWithFillLimits {
+    type Err = Bla;
+
+    fn from_str(s: &str) -> Result<Self, Bla> {
```

Comment:
Is it possible to split this looong function into 4-5 smaller ones? naively it seems there is a clear path step 1 -> step 2 -> step 3 -> ... which act in isolation

Reply:
I can split this function into smaller ones, but I don't see that making it more readable. But let me try to improve it.

Reply:
to me it would be more readable - because then there will be a clear input, a clear task, and a clear output. Needless to say that this would also be good for unit testing.

Reply:
I agree that if this could be broken into smaller functions that'd be great as right now it is a bit difficult to read.